pilot/pkg/proxy/envoy/v1/watcher.go

```diff
@@ -208,6 +222,54 @@ func watchCerts(ctx context.Context, certsDirs []string, watchFileEventsFn watch
 	watchFileEventsFn(ctx, fw.Event, minDelay, updateFunc)
 }
 
+// waitForCertsPresent is a blocking function that periodically checks the existance
+// of the certs, once every interval. It returns nil when all cert files are non-empty.
+// It returns error if the timeout is reached.
+func waitForCertsPresent(certs []CertSource, interval time.Duration, timeout time.Duration) (err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		for _, cert := range certs {
+			// Check the cert file directories one by one.
+			for {
+				err := checkCerts(cert)
+				if err == nil {
+					break
+				}
+				log.Warnf("%v. Will retry in %v", err, interval)
+				select {
+				case <-time.After(interval):
+					// retry
+				case <-ctx.Done():
+					return fmt.Errorf("certs do not present after timeout %v", timeout)
+				}
+			}
+		}
+		return nil
+	})
+	err = g.Wait()
+	cancel()
```

Comment:
As a general practice its probably better to use `defer cancel()` immediately after `context.WithTimeout` to avoid leaks.

Reply:
... and then you could use `return g.Wait()`